exchange/exchange.go

```diff
@@ -162,12 +163,95 @@ func (e *exchange) HoldAuction(ctx context.Context, bidRequest *openrtb.BidReque
 			}
 			targData.setTargeting(auc, bidRequest.App != nil, bidCategory)
 		}
+
+		if requestExt.Prebid.SupportDeals {
+			applyDealSupport(bidRequest, auc)
+		}
 	}
 
 	// Build the response
 	return e.buildBidResponse(ctx, liveAdapters, adapterBids, bidRequest, resolvedRequest, adapterExtra, auc, errs)
 }
 
+type DealTierInfo struct {
+	Prefix      string `json:"prefix"`
+	MinDealTier int    `json:"minDealTier"`
+}
+
+type DealTier struct {
+	Info *DealTierInfo `json:"dealTier,omitempty"`
+}
+
+type BidderDealTier struct {
+	DealInfo map[string]*DealTier
+}
+
+// Updates targeting keys with deal prefixes if minimum deal tier exceeded
+func applyDealSupport(bidRequest *openrtb.BidRequest, auc *auction) {
+	impDealMap := getDealTiers(bidRequest)
+
+	for impID, topBidsPerImp := range auc.winningBidsByBidder {
+		impDeal := impDealMap[impID].DealInfo
+		for bidderName, topBidPerBidder := range topBidsPerImp {
+			bidderString := bidderName.String()
+
+			if validateDealTier(impDeal[bidderString]) {
+				updateCatDur(topBidPerBidder, impDeal[bidderString].Info)
+			}
+		}
+	}
+}
+
+// Creates map of impression to bidder deal tier configuration
+func getDealTiers(bidRequest *openrtb.BidRequest) map[string]*BidderDealTier {
+	impDealMap := make(map[string]*BidderDealTier)
+
+	for _, imp := range bidRequest.Imp {
+		var bidderDealTier BidderDealTier
+		err := json.Unmarshal(imp.Ext, &bidderDealTier.DealInfo)
+		if err != nil {
+			continue
+		}
+
+		impDealMap[imp.ID] = &bidderDealTier
+	}
+
+	return impDealMap
+}
+
+func validateDealTier(impDeal *DealTier) bool {
+	return impDeal.Info != nil && len(impDeal.Info.Prefix) > 0
+}
+
+func updateCatDur(topBidPerBidder *pbsOrtbBid, dealTierInfo *DealTierInfo) {
+	var anExt AppNexusExt
```

Comment:
We shouldn't be accessing AppNexus specific extension in the exchange.go file. Rather than parsing the appnexus extension here, the generic "prebid" extension should have a field for the deal priority that is used here, and that field needs to be set by the adapter for any bids it returns that have a deal priority.

Reply:
Adding in an updated to fix this. This type here is now removed and the deal priority is set by the adapter instead from the TypedBid.